### PR TITLE
os/bluestore: fix deep-scrub operation againest disk silent errors

### PR DESCRIPTION
**Pull request by wangxiaoguang**
Say a object who has data caches, but in a while later, caches' underlying
physical device has silent disk erros accidentally, then caches and physical
data are not same. In such case, deep-scrub operation still tries to read
caches firstly and won't do crc checksum, then deep-scrub won't find such
data corruptions timely.

Here introduce a new flag 'CEPH_OSD_OP_FLAG_BYPASS_CACHE' which tells
deep-scrub to bypass object caches. Note that we only bypass cache who is in
STATE_CLEAN state. For STATE_WRITING caches, currently they are not written
to physical device, so deep-scrub operation can not read physical device and
can read these dirty caches safely. Once they are in STATE_CLEAN state(or not
added to bluestore cache), next round deep-scurb can check them correctly.

As to above discussions, I refactor BlueStore::BufferSpace::read sightly,
adding a new 'flags' argument, whose value will be:
```c++
     enum {
       READ_CLEAN_CACHE = 0x1,     // read clean cache
       READ_DIRTY_CACHE = 0x2,     // read dirty cache
       READ_ALL_CACHE   = 0x3,     // read clean & dirty cache
    };
```
flags READ_ALL_CACHE: normal read
flags READ_DIRTY_CACHE: bypass clean cache, currently only for deep-scrube
                        operation

Test:
   I deliberately corrupt a object with cache, with this patch, deep-scrub
   can find data error very timely.

Signed-off-by: Xiaoguang Wang <xiaoguang.wang@easystack.cn>

### CODE REVIEW

**Comment by wangxiaoguang** [comment]
@liewegas  please take a look at this patch, thanks.

**Comment by liewegas** [comment]
This is great!  I would modify it slightly, though, so that there is a single BYPASS_CLEAN_CACHE flag, because it doesn't make sense to use clean and not dirty buffers (READ_CLEAN_CACHE).

One question: did check what the behavior was when you have a clean buffer, bypass it and read the device but get an error? is the clean buffer still there so that the next (normal) read will succeed? In principle there is an opportunity to recover some fo the data (write it somewhere else etc) although in practice I wouldn't worry about this narrow corner case too much, at least not for a first pass.

**Comment by wangxiaoguang** [comment]
> This is great! I would modify it slightly, though, so that there is a single BYPASS_CLEAN_CACHE flag, because it doesn't make sense to use clean and not dirty buffers (READ_CLEAN_CACHE).

ok, thanks for review:)

> did check what the behavior was when you have a clean buffer, bypass it and read the device but get an error?
This object will be repaired.
  ==>> do_osd_ops
  ====>> do_read
  ======>> objects_read_sync // get EIO error.
  ======>> rep_repair_primary_object 
then read operation will be issued again. I use  “rados get” to test this case, 'rados' tool does not perceive this object's corruption even.

> is the clean buffer still there so that the next (normal) read will succeed?
Yes, clean buffer is there and next read will succeed(indeed current read will also succeed, corrupted object will be fix immediately), but I'm not sure whether caches were discarded and re-generated in the procedure of fixing object.

File: src/include/rados.h [hunk 1]
```diff
@@ -470,6 +470,7 @@ enum {
 	CEPH_OSD_OP_FLAG_FADVISE_DONTNEED   = 0x20,/* data will not be accessed in the near future */
 	CEPH_OSD_OP_FLAG_FADVISE_NOCACHE   = 0x40, /* data will be accessed only once by this client */
 	CEPH_OSD_OP_FLAG_WITH_REFERENCE   = 0x80, /* need reference couting */
+	CEPH_OSD_OP_FLAG_BYPASS_CLEAN_CACHE = 0x100, /* bypass ObjectStore cache, mainly for deep-scrub */
```
Comment (xiexingguo) [inline src/include/rados.h]: need to update `ceph_osd_op_flag_name` too:
```c++
const char * ceph_osd_op_flag_name(unsigned flag)
{
  const char *name;

  switch(flag) {
    case CEPH_OSD_OP_FLAG_EXCL:
      name = "excl";
      break;
    case CEPH_OSD_OP_FLAG_FAILOK:
      name = "failok";
      break;
    case CEPH_OSD_OP_FLAG_FADVISE_RANDOM:
      name = "fadvise_random";
      break;
    case CEPH_OSD_OP_FLAG_FADVISE_SEQUENTIAL:
      name = "fadvise_sequential";
      break;
    case CEPH_OSD_OP_FLAG_FADVISE_WILLNEED:
      name = "favise_willneed";
      break;
    case CEPH_OSD_OP_FLAG_FADVISE_DONTNEED:
      name = "fadvise_dontneed";
      break;
    case CEPH_OSD_OP_FLAG_FADVISE_NOCACHE:
      name = "fadvise_nocache";
      break;
    default:
      name = "???";
  };

  return name;
}
```

Comment (wangxiaoguang) [reply src/include/rados.h]: Thanks, I'll update this patch soon.

File: src/osd/osd_types.cc [hunk 2]
```diff
@@ -100,6 +100,9 @@ const char * ceph_osd_op_flag_name(unsigned flag)
     case CEPH_OSD_OP_FLAG_FADVISE_NOCACHE:
       name = "fadvise_nocache";
       break;
+    case CEPH_OSD_OP_FLAG_BYPASS_CLEAN_CACHE:
```
Comment (xiexingguo) [inline src/osd/osd_types.cc]: Do you mind adding the new **CEPH_OSD_OP_FLAG_WITH_REFERENCE** flag too?

Comment (wangxiaoguang) [reply src/osd/osd_types.cc]: ok, will add it.

File: src/osd/osd_types.cc [hunk 3]
```diff
@@ -100,6 +100,12 @@ const char * ceph_osd_op_flag_name(unsigned flag)
     case CEPH_OSD_OP_FLAG_FADVISE_NOCACHE:
       name = "fadvise_nocache";
       break;
+    case CEPH_OSD_OP_FLAG_WITH_REFERENCE:
+      name = "reference_couting";
```
Comment (xiexingguo) [inline src/osd/osd_types.cc]: Don't want to be nit-picking, but I'd prefer `with_reference `to keep pace with others...

Comment (wangxiaoguang) [reply src/osd/osd_types.cc]: Really never mind, your suggestions are good and reasonable, indeed I should appreciate your patience! It's me not doing well, thanks :100:

File: src/osd/osd_types.cc [hunk 4]
```diff
@@ -100,6 +100,12 @@ const char * ceph_osd_op_flag_name(unsigned flag)
     case CEPH_OSD_OP_FLAG_FADVISE_NOCACHE:
       name = "fadvise_nocache";
       break;
+    case CEPH_OSD_OP_FLAG_WITH_REFERENCE:
+      name = "with_reference";
+      break;
```
Comment (dzafman) [inline src/osd/osd_types.cc]: These 3 lines should be in a separate commit, in case it gets backported and this pull request doesn't.

Comment (wangxiaoguang) [reply src/osd/osd_types.cc]: It's reasonable,  thanks! New version will be sent soon.

**Comment by wangxiaoguang** [comment]
@dzafman  @xiexingguo  @tchaikov  I submitted new version which just divided previous patch into twos, thanks.

**Comment by dzafman** [comment]
retest this please

File: src/os/bluestore/BlueStore.cc [hunk 5]
```diff
@@ -7442,6 +7450,13 @@ int BlueStore::_do_read(
 
   ready_regions_t ready_regions;
 
+  // for deep-scrub, wo only read dirty cache and bypass clean cache in
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc]: s/wo/we/

File: src/os/bluestore/BlueStore.cc [hunk 6]
```diff
@@ -7442,6 +7450,13 @@ int BlueStore::_do_read(
...
+  // order to read undering block device in case there are silent disk errors.
```
Comment (tchaikov) [inline src/os/bluestore/BlueStore.cc]: underlying.

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-08-30_02:23:40-rados-wip-kefu-testing-2018-08-29-2346-distro-basic-smithi/
the failures are tracked by 

- http://tracker.ceph.com/issues/24485
- http://tracker.ceph.com/issues/34529

hence i don't think they are relevant.

**Comment by tchaikov** [comment]
@liewegas @dzafman @xiexingguo shall we backport this change?

**Comment by liewegas** [comment]
I think so!

File: src/os/bluestore/BlueStore.cc (hunk 6 above)
Comment (wangxiaoguang) [reply src/os/bluestore/BlueStore.cc]: Sorry, will send new version.

Comment (xiexingguo) [reply src/os/bluestore/BlueStore.cc]: s/wo/we/ ?

Comment (xiexingguo) [reply src/os/bluestore/BlueStore.cc]: Since you are there... :-)

**Comment by tchaikov** [comment]
http://tracker.ceph.com/issues/34541
